cpu/qn908x/periph/uart.c

```diff
@@ -0,0 +1,293 @@
+        dev->FIFOCFG |= USART_FIFOCFG_EMPTYRX_MASK |
+                        USART_FIFOCFG_ENABLERX_MASK;
+        dev->FIFOINTENSET = USART_FIFOINTENSET_RXLVL_MASK;
+        /* flexcomm_num is the same as the USART instance number in the
+         * USART_IRQS array. */
+        const uint8_t usart_irqn[] = USART_IRQS;
+        NVIC_EnableIRQ(usart_irqn[flexcomm_num]);
+    }
+    else {
+        dev->FIFOCFG &= ~USART_FIFOCFG_ENABLERX_MASK;
+        uart_conf->dev->FIFOINTENCLR = USART_FIFOINTENSET_RXLVL_MASK;
+    }
+
+    /* Enable TX side. */
+    if (uart_conf->tx_pin != GPIO_UNDEF) {
+        dev->FIFOCFG |= USART_FIFOCFG_EMPTYTX_MASK |
+                        USART_FIFOCFG_ENABLETX_MASK;
+    }
+    else {
+        dev->FIFOCFG &= ~USART_FIFOCFG_ENABLETX_MASK;
+    }
+
+    /* Configure RX and TX pins. RX/TX pins are always in function 4.
+     * GPIO_UNDEF are ignored. */
+    gpio_init_mux(uart_conf->rx_pin, 4);
+    gpio_init_mux(uart_conf->tx_pin, 4);
+    /* This call also enables the UART. */
+    return uart_mode(uart, UART_DATA_BITS_8, UART_PARITY_NONE,
+                     UART_STOP_BITS_1);
+}
+
+int uart_mode(uart_t uart, uart_data_bits_t data_bits, uart_parity_t parity,
+              uart_stop_bits_t stop_bits)
+{
+    /* Setup mode and enable USART. The values of the uart_data_bits_t,
+     * uart_parity_t and uart_stop_bits_t enums were selected to match the
+     * fields in this registers so there's no need to do any conversion. */
+    uart_config[uart].dev->CFG = USART_CFG_PARITYSEL(parity)
+                                 | USART_CFG_STOPLEN(stop_bits)
+                                 | USART_CFG_DATALEN(data_bits)
+                                 | USART_CFG_LOOP(0) | USART_CFG_ENABLE_MASK;
+    return UART_OK;
+}
+
+void uart_write(uart_t uart, const uint8_t *data, size_t len)
+{
+    USART_Type *dev = uart_config[uart].dev;
+
+    /* If the TX side or the whole uart mode was not enabled during init or at
+     * all yet we can only ignore this transmission. This allows DEBUG messages
+     * to be ignored without hanging here before the uart is initialized. */
+    if (!(dev->FIFOCFG & USART_FIFOCFG_ENABLETX_MASK) ||
+        !(dev->CFG & USART_CFG_ENABLE_MASK)) {
+        return;
+    }
+
+    for (; len; len--) {
+        while (!(dev->FIFOSTAT & USART_FIFOSTAT_TXNOTFULL_MASK)) {}
+        dev->FIFOWR = *(data++);
+    }
+    /* Wait until we flush out all the bytes. */
+    while (!(dev->STAT & USART_STAT_TXIDLE_MASK)) {}
+}
+
+void uart_poweron(uart_t uart)
+{
+    USART_Type *dev = uart_config[uart].dev;
+
+    dev->CFG |= USART_CFG_ENABLE_MASK;
+}
+
+void uart_poweroff(uart_t uart)
+{
+    USART_Type *dev = uart_config[uart].dev;
+
+    while (!(dev->STAT & USART_STAT_TXIDLE_MASK)) {}
+    dev->CFG &= ~USART_CFG_ENABLE_MASK;
+}
+
+void isr_flexcomm_uart(USART_Type *dev, uint32_t flexcomm_num)
+{
+    uart_t uart = uart_dev_from_flexcomm[flexcomm_num];
+
+    for (uint32_t flags = dev->FIFOSTAT;
```

Comment:
why not just 

```suggestion
    while (dev->FIFOSTAT & USART_FIFOSTAT_RXNOTEMPTY_MASK) {
```

Reply:
I don't know why not, maybe when I wrote this I had an use for `flags`. I'll update it.